netket/utils/static_range.py

```diff
@@ -149,12 +169,14 @@ def states_to_numbers(self, x, dtype: DType = int):
         Returns:
             An array of integers, which can be.
         """
-        idx = (x - self.start) / self.step
-        if dtype is not None:
-            if not hasattr(idx, "astype"):
-                idx = np.array(idx, dtype=dtype)
-            else:
-                idx = idx.astype(dtype)
+        idx = (x - self.start) // self.step
+        if dtype is None:
+            dtype = self.dtype if np.issubdtype(self.dtype, np.integer) else int
```

Comment:
This is wrong.

The dtype of `StaticRange` is the dtype of the values in the range, not the dtype of the integer used to enumerate them...

Reply:
I was thinking of adding an additional dtype for numbers. The best would be an unsigned integer.
However, I was looking at it from the memory perspective and code simplicity and, I assume, speed:

- `start`/`stop` are integers:
Then length is at most `end-start` since the minimal step is 1, i.e. it is at most  `2*np.iinfo(np.intX).max`. 
In the case of `StaticRange(start=-127, step=1, length=254)` we have to first cast states into `int16` from `int8` otherwise we get negative numbers. Now, one could cast negative numbers to `uint8` and still be correct, but the problem with esoteric case with `step=3` above remains (which can be solved with proper calculation with modulo mentioned above).
This is why, I also test that `length * step` is in the range of dtype. Yes, we lose 1/2 of dtype range for `start-end`, but I don't think there are many cases with spin or boson number larger than 63/2, and then we can just go to `int16`.
- if start/stop are floats, the memory usage is already high, and the difference int64 between int8 shouldn't have a large effect.

Reply:
> I was thinking of adding an additional dtype for numbers. The best would be an unsigned integer.

I'm against. It's adding complexity with no added benefit that I understand.
the `to_number` function takes a dtype, so if it's specified we'd be ignoring the 'intrinsic' one anyway.

What we should do is have a better, smarter default.
The better, smarter default could just be the first unsigned integer dtype to pass `_in_int_dtype_range(StaticRange.length, dtyp)`. 

But this would be a default, not something stored in the object.

--
> This is why, I also test that length * step is in the range of dtype. Yes, we lose 1/2 of dtype range for start-end, but I don't think there are many cases with spin or boson number larger than 63/2, and then we can just go to int16.

Ok, good point. Just change the code to compute the bottom type .

> if start/stop are floats, the memory usage is already high, and the difference int64 between int8 shouldn't have a large effect.

Let's assume they are not floats.
--

The reason for avoiding adding new fields is that it adds complexity for what I believe is no really useful feature. This stuff is quite internal, I don't expect users to directly interact with creating StaticRanges often...